Stop using WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/crypto/

cdumez opened the pull request:
#### 490f610d13ba000918520c7bb501a7c458b93142
<pre>
Stop using WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/crypto/
<a href="https://bugs.webkit.org/show_bug.cgi?id=284018">https://bugs.webkit.org/show_bug.cgi?id=284018</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/CryptographicUtilities.cpp:
(WTF::constantTimeMemcmp):
* Source/WTF/wtf/CryptographicUtilities.h:
(WTF::constantTimeMemcmp):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCTR.cpp:
(WebCore::parametersAreValid):
(WebCore::CryptoAlgorithmAESCTR::CounterBlockHelper::CounterBlockHelper):
(WebCore::CryptoAlgorithmAESCTR::CounterBlockHelper::counterVectorAfterOverflow const):
(WebCore::CryptoAlgorithmAESCTR::CounterBlockHelper::CounterBlockBits::set):
(WebCore::CryptoAlgorithmAESCTR::CounterBlockHelper::CounterBlockBits::all const):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESGCM.cpp:
(WebCore::tagLengthIsValid):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
(WebCore::CryptoAlgorithmX25519::deriveBits):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESCBCMac.cpp:
(WebCore::transformAESCBC):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESCFBMac.cpp:
(WebCore::transformAESCFB):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
(WebCore::decyptAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHMACMac.cpp:
(WebCore::platformVerifyCC):
* Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
* Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
* Source/WebCore/crypto/cocoa/CryptoKeyRSAMac.cpp:
(WebCore::CryptoKeyRSA::importSpki):
(WebCore::CryptoKeyRSA::importPkcs8):
* Source/WebCore/crypto/cocoa/CryptoUtilitiesCocoa.cpp:
(WebCore::transformAESCTR):
* Source/WebCore/crypto/cocoa/SerializedCryptoKeyWrapMac.mm:
(WebCore::unwrapCryptoKey):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmAESGCMGCrypt.cpp:
(WebCore::gcryptDecrypt):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmHMACGCrypt.cpp:
(WebCore::CryptoAlgorithmHMAC::platformVerify):
* Source/WebCore/crypto/openssl/CryptoAlgorithmHMACOpenSSL.cpp:
(WebCore::CryptoAlgorithmHMAC::platformVerify):

Canonical link: <a href="https://commits.webkit.org/287365@main">https://commits.webkit.org/287365@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81da59be51095a654e8cc28d171b54a824d054dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79336 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32707 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83951 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30485 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81469 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6626 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62055 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19952 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82403 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52121 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72286 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42362 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49465 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26468 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28883 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72431 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70578 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26913 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85351 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78533 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6620 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4604 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70305 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6784 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68160 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69552 "Found 6 new API test failures: /TestWebKit:WebKit.DownloadDecideDestinationCrash, /TestWebKit:WebKit.FrameMIMETypePNG, /TestWebKit:WebKit.EvaluateJavaScriptThatThrowsAnException, /TestWebKit:WebKit.GeolocationBasicWithHighAccuracy, /TestWebKit:WebKit2UserMessageRoundTripTest.WKString, /TestWebKit:WebKit.ParentFrame (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13607 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12469 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100867 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12251 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6572 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24609 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6491 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9964 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8287 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmAESGCM.cpp:
```diff
@@ -54,11 +53,7 @@ static inline bool usagesAreInvalidForCryptoAlgorithmAESGCM(CryptoKeyUsageBitmap
-            return true;
-    }
-    return false;
+    return std::find(validTagLengths.begin(), validTagLengths.end(), tagLength) != validTagLengths.end();
```
More modern way to write this is `std::ranges::contains(validTagLengths, tagLength)`, but not sure we can use that yet. It’s C++20, for what it’s worth.

cdumez replied:
I'll try and see if the bots are happy. Thanks!

cdumez replied:
Sadly, some of the bots didn't like it:
```
In file included from /Volumes/Data/worker/macOS-Ventura-Release-Build-EWS/build/WebKitBuild/Release/DerivedSources/WebCore/unified-sources/UnifiedSource118.cpp:2:
/Volumes/Data/worker/macOS-Ventura-Release-Build-EWS/build/Source/WebCore/crypto/algorithms/CryptoAlgorithmAESGCM.cpp:57:25: error: no member named 'contains' in namespace 'std::ranges'
```

darinadler replied:
We can give up on it of course unless it's just a missing include.

cdumez replied:
Oh, I had includes `<ranges>` but it seems I need `<algorithm>` for std::ranges::contains. I'll try.

cdumez replied:
Actually, both `<ranges>` and `<algorithm>` were included so I don't think that was the issue. I'm trying with `std::ranges::find()` now as a fall back.

cdumez replied:
`std::ranges::find()` looks good on the bots so far.

cdumez committed:
Stop using WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/crypto/
https://bugs.webkit.org/show_bug.cgi?id=284018

Reviewed by Darin Adler.

* Source/WTF/wtf/CryptographicUtilities.cpp:
(WTF::constantTimeMemcmp):
* Source/WTF/wtf/CryptographicUtilities.h:
(WTF::constantTimeMemcmp):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCTR.cpp:
(WebCore::parametersAreValid):
(WebCore::CryptoAlgorithmAESCTR::CounterBlockHelper::CounterBlockHelper):
(WebCore::CryptoAlgorithmAESCTR::CounterBlockHelper::counterVectorAfterOverflow const):
(WebCore::CryptoAlgorithmAESCTR::CounterBlockHelper::CounterBlockBits::set):
(WebCore::CryptoAlgorithmAESCTR::CounterBlockHelper::CounterBlockBits::all const):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESGCM.cpp:
(WebCore::tagLengthIsValid):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
(WebCore::CryptoAlgorithmX25519::deriveBits):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESCBCMac.cpp:
(WebCore::transformAESCBC):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESCFBMac.cpp:
(WebCore::transformAESCFB):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
(WebCore::decyptAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHMACMac.cpp:
(WebCore::platformVerifyCC):
* Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
* Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
* Source/WebCore/crypto/cocoa/CryptoKeyRSAMac.cpp:
(WebCore::CryptoKeyRSA::importSpki):
(WebCore::CryptoKeyRSA::importPkcs8):
* Source/WebCore/crypto/cocoa/CryptoUtilitiesCocoa.cpp:
(WebCore::transformAESCTR):
* Source/WebCore/crypto/cocoa/SerializedCryptoKeyWrapMac.mm:
(WebCore::unwrapCryptoKey):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmAESGCMGCrypt.cpp:
(WebCore::gcryptDecrypt):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmHMACGCrypt.cpp:
(WebCore::CryptoAlgorithmHMAC::platformVerify):
* Source/WebCore/crypto/openssl/CryptoAlgorithmHMACOpenSSL.cpp:
(WebCore::CryptoAlgorithmHMAC::platformVerify):

Canonical link: https://commits.webkit.org/287365@main
